the-collab-lab/tcl-40-smart-shopping-list
# sp-dh-new-shopping-list

_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

The code in this PR aims to do the following:
 - Upon loading, the Home component checks to see if localStorage contains a key named 'token' associated with the User. 
 - If there is a token, the JoinList component is rendered:

-- Once rendered,

    -- The value of the existing token will be shown as the default value in the input field
    -- Render the "List" view

- If there isn't a token, the CreateList component is rendered:

-- Once rendered, a button is clicked to:

    -- Generate a new token
    -- Save the new token to localStorage
    -- Renders the "List" view

We created two components (CreateList and JoinList) in order to simplify the Home component's return function as well as making them easy to reuse if necessary. 

## Related Issue

closes #3



## Acceptance Criteria



- @the-collab-lab/shopping-list-utils has been added as a dependency to the project.
- For users who do not already have a token or a list, a button or link exists on the home screen that allows them to create a new list
- Clicking the button/link generates a new token and saves it to localStorage
- Once the token has been created and saved, the user is redirected to the "list" view
- For users who do have a token already saved in localStorage, they should be shown/redirected to the "list" view

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before




### After

 - CreateList (User does not yet have a token)
<img width="617" alt="Screen Shot 2022-04-13 at 7 59 08 PM" src="https://user-images.githubusercontent.com/52385888/163288730-288270a2-982e-43a5-b2b2-ce96c766bb72.png">

- JoinList (if User already has a token)

<img width="493" alt="Screen Shot 2022-04-13 at 9 16 01 PM" src="https://user-images.githubusercontent.com/52385888/163295175-ccc4b518-a584-45fe-839c-8f568ce8f9fd.png">



## Testing Steps / QA Criteria



In the App:

- Pull down the changes from the branch
`git pull`

- Checkout to our branch
`git checkout sj-dh-new-shopping-list`

- Run server
`npm start`

- Open `localhost:3000` and open your console

- Navigate to the "Application" tab and on the left under the Storage heading, click on "Local Storage". 
![Screenshot 2022-04-13 201434](https://user-images.githubusercontent.com/80491609/163306576-76471de2-532f-4cf1-9bfd-3f5c3990008f.png)

- There should be nothing under "Key" and "Value" (ignore the "ally-supports-cache" in the screenshot above)

- Click on the "Join an existing list" button while the Application window is still open

- After you click on the button there should be a new Key of "token" and a Value of 3 randomly generated words in the Application window
![Animation](https://user-images.githubusercontent.com/80491609/163307092-85902cdd-0eab-4a4d-813e-f80ccc2b5b5d.gif)

- Click on the "Home" link. You should then see an input field that has already been filled out with the same 3 words from the Value window
![Animation1](https://user-images.githubusercontent.com/80491609/163307443-e032e99b-0e24-4425-a3c4-2a2a79982d36.gif)

- If you want to retest, right click on the token inside your Local Storage and click "Delete" and then refresh and click on the button to generate a new token.
![Animation2](https://user-images.githubusercontent.com/80491609/163307819-6d14324a-a2e5-48db-b78a-05d4d94a4a2e.gif)
